src/Interpreters/InterpreterCreateQuery.cpp

```diff
@@ -1712,7 +1734,12 @@ AccessRightsElements InterpreterCreateQuery::getRequiredAccess() const
     else
     {
         if (create.temporary)
-            required_access.emplace_back(AccessType::CREATE_TEMPORARY_TABLE);
+        {
+            if (create.storage && create.storage->engine && create.storage->engine->name != "Memory")
+                required_access.emplace_back(AccessType::CREATE_ARBITRARY_TEMPORARY_TABLE);
+            else
+                required_access.emplace_back(AccessType::CREATE_TEMPORARY_TABLE);
```

Comment:
It will stop working once support for `default_table_engine` is added for temporary tables. It's okay, but let's add a comment that we assume that `default_table_engine` does not affect temporary tables

Reply:
What text should we use the comment? For example:

default_table_engine does not affect temporary tables

OR

Currently default table engine for temporary tables is Memory. default_table_engine does not affect temporary tables.

Reply:
both are okay

Reply:
OK. I added the comment.

Reply:
> It will stop working once support for `default_table_engine` is added for temporary tables.

I was thinking about the usage of  `default_table_engine`  for  temporary tables in future PRs. 
My point of view: for temporary tables we should not use use  `default_table_engine` but we should implement new parameter  `default_table_engine_for_temporary_tables` (or whatever better name) with default value `Memory`. This is because the temporary tables are the special type of entities and it deserves a separate parameter. 

So I think `MergeTree` is good for  `default_table_engine` (while I see that default value is [None](https://clickhouse.com/docs/en/operations/settings/settings/#default_table_engine) that's right).  But  `Memory` is good for `default_table_engine_for_temporary_tables`.

The second related idea:
If we use external data `--external` we may set name of temporary table with `--name` but how can we set the engine? Probably it maybe done by setting `default_table_engine_for_temporary_tables` . Or (second variant) there should be new parameter like `--engine`
In this case if we decide to use `MergeTree` (because external data is big in size) how can we specify ORDER BY/PRIMARY KEY? Because it's required by `MergeTree` engines.

These are my two thoughts about future features/options for temporary tables.